### PR TITLE
Added children to context bound object

### PR DESCRIPTION
**Pull request by amitapl**
Please take a look as it may have things you may not like.

### CODE REVIEW

File: Story.Core/ContextBoundObject.cs [hunk 1]
```diff
@@ -73,6 +89,7 @@ protected void Attach()
                 if (existingContext != null)
                 {
                     this.Parent = existingContext;
+                    this.Parent._childrenBag.Add(this as T);
```
Comment (israelchen) [inline Story.Core/ContextBoundObject.cs]: don't conditionally cast here.

File: Story.Core/Story.cs [hunk 2]
```diff
@@ -8,7 +9,7 @@
     using Utils;
 
     [Serializable]
-    public class Story : ContextBoundObject<IStory>, IStory
+    public class Story : ContextBoundObject<Story>, IStory
```
Comment (israelchen) [inline Story.Core/Story.cs]: Why the change here? It imposes a restriction that children/parent must all derive from Story. Do we want that?

File: Story.Core/StoryExtensions.cs [hunk 3]
```diff
@@ -141,9 +149,147 @@ public static void Error(this IStory story, string format, params object[] args)
...
+        /// Get the first value from the story data in the story
+        /// and the children stories inside (recursive)
+        /// or null if not found
+        /// </summary>
+        public static object GetDataValue(this IStory story, string key, bool recursive = true)
+        {
+            Ensure.ArgumentNotNull(story, "story");
+
+            // assume story.Data[key] returns null if not found
+            object result = story.Data[key];
```
Comment (israelchen) [inline Story.Core/StoryExtensions.cs]: Shouldn't this be GetData() now?

File: Story.Core/StoryExtensions.cs [hunk 4]
```diff
@@ -141,9 +149,147 @@ public static void Error(this IStory story, string format, params object[] args)
...
+            if (result != null || recursive == false)
```
Comment (israelchen) [inline Story.Core/StoryExtensions.cs]: actually, we have a problem with nulls i think. It is perfectly fine to save a null object value since it could
indicate a really production issue. e.g.: userId was null and that's why we failed later on etc.

we need to differentiate between "value is null" and "key does not exist".

File: Story.Core/StoryExtensions.cs [hunk 5]
```diff
@@ -141,9 +149,147 @@ public static void Error(this IStory story, string format, params object[] args)
...
+
+            if (recursive == false)
+            {
+                return story.Log;
+            }
+
+            return story.Log.Union(story.Children.Flatten(childStory => childStory.Children).SelectMany(childStory => childStory.Log));
+        }
+
+        public static IEnumerable<T> Flatten<T>(this IEnumerable<T> source, Func<T, IEnumerable<T>> childrenSelector)
```
Comment (israelchen) [inline Story.Core/StoryExtensions.cs]: validate source, childrenSelector are not null

File: Story.Core/ContextBoundObject.cs (hunk 1 above)
Comment (amitapl) [inline Story.Core/ContextBoundObject.cs]: yep

File: Story.Core/Story.cs (hunk 2 above)
Comment (amitapl) [inline Story.Core/Story.cs]: I had to do it, but I am open to a way to fix this whole ContextBoundObject<T>

The original issue was casting from IEnumerable<ContextBoundObject<T>> to IEnumerable<IStory>

File: Story.Core/StoryExtensions.cs (hunk 3 above)
Comment (amitapl) [inline Story.Core/StoryExtensions.cs]: not really, GetData returns all data from all children
this is for a single value
